File: mpbeps2/ReadFiles/preadf2.f90
!-----------------------------------------------------------------------
! This program reads real periodic 2d scalar data
! written for 2D MPI/OpenMP PIC codes
! written by Viktor K. Decyk, UCLA
      program preadf2
      use in2, only: idrun, indx, indy, nvp, tend, dt, ci,              &
     &pot2d, ntp, fpname, modesxp, modesyp, nprec,                      &
     &dene2d, ntde, fdename, modesxde, modesyde, nderec,                &
     &deni2d, ntdi, fdiname, modesxdi, modesydi, ndirec
      use graf2
!
      implicit none
      integer, parameter :: ns = 3
      integer :: iudm = 19, ius = 11
      integer :: i, n, m, nx, ny, kyp, kyb, nyv, lrec, nrec, ios, ierr
      integer :: nplot = 1
      integer, dimension(ns) :: nscalars = 0
      real, dimension(:,:), allocatable :: sfield
      character(len=16), dimension(ns) :: dname = (/'potential       ', &
     &'electron density','ion density     '/)
      character(len=10) :: cdrun
      character(len=32) :: fname
!
      write (*,*) 'enter idrun:'
      read (5,*) idrun
      write (cdrun,'(i10)') idrun
      cdrun = adjustl(cdrun)
      fname = 'diag2.'//cdrun
      open(unit=iudm,file=fname,form='formatted',status='old')
!
! determine which scalar diagnostics are available
      do n = 1, ns
      select case(n)
! load metadata for potential data
      case (1)
         read (iudm,pot2d,iostat=ios)
! load metadata for electron density data
      case (2)
         read (iudm,dene2d,iostat=ios)
! load metadata for ion density data
      case (3)
         read (iudm,deni2d,iostat=ios)
      end select
      if (ios==0) nscalars(n) = 1
      rewind iudm
      enddo
!
! select diagnostic
      m = sum(nscalars)
      if (m > 1) then
         n = -1
         do while (n < 0)
            do i = 1, ns
               if (nscalars(i)==1) then
                  write (*,*) 'enter ', i, 'for ', trim(dname(i))
               endif
            enddo
            read (5,*) n
            if (n==0) stop
            if ((n >= 1).and.(n <= ns)) then
               if (nscalars(n)==0) n = -1
            else
               n = -1
            endif
            if (n < 0) then
               write (*,*) 'invalid entry, try again or enter 0 to quit'
            endif
         enddo
      else if (m==1) then
         do i = 1, ns
            if (nscalars(i)==1) then
               n = i
               exit
            endif
         enddo
      else
         write (*,*) 'no scalar diagnostic files found'
         stop
      endif
!
      write (*,*) trim(dname(n)), ' diagnostic selected'
!
      select case(n)
      case (1)
         read (iudm,pot2d,iostat=ios)
         fname = fpname; nrec = nprec
      case (2)
         read (iudm,dene2d,iostat=ios)
         fname = fdename; nrec = nderec
      case (3)
         read (iudm,deni2d,iostat=ios)
         fname = fdiname; nrec = ndirec
      end select
!
! nx/ny = number of global grid points in x/y direction
      nx = 2**indx; ny = 2**indy
! kyp = number of real grids in each field partition in y direction
      kyp = (ny - 1)/nvp + 1
! kyb = minimum number of processors in distributed array
      kyb = (ny - 1)/kyp + 1
! nyv = second dimension of scalar field array, >= ny
      nyv = kyp*kyb
!
! allocate scalar array
      allocate(sfield(nx,nyv))
! open direct access file for scalar field
      inquire(iolength=lrec) sfield(1,1); lrec = lrec*nx*nyv
      open(unit=ius,file=fname,form='unformatted',access='direct',      &
     &recl=lrec,status='old')
!
! nrec = number of complete records
      nrec = nrec/kyb
      write (*,*) 'records found: nrec = ', nrec
!
! open graphics device
      ierr = open_graphs(nplot)
! set palette to color wheel
      call STPALIT(2)
!
! read and display data
      do i = 1, nrec
         read (unit=ius,rec=i) sfield
         call dscaler2(sfield,trim(dname(n)),i,999,0,1,nx,ny,ierr)
         if (ierr==1) exit
      enddo
!
      call close_graphs
!
      end program
!
! unneeded function in input2mod.f90
      subroutine PPBDCAST(f,nxp)
      implicit none
      integer, intent(in) :: nxp
      double precision, dimension(nxp), intent(inout) :: f
      end subroutine
!

File: mpbeps2/ReadFiles/preadf2.py
#-----------------------------------------------------------------------
# This program reads real periodic 2d scalar data
# written for 2D MPI/OpenMP PIC codes
# written by Viktor K. Decyk, UCLA
from __future__ import print_function
import sys
import math
import numpy

sys.path.append('./ReadFiles')
from fgraf import *
from libcmfield2 import *

if (sys.version_info.major==2):
   input = raw_input

#-----------------------------------------------------------------------
def main(idrun):
   """
   reads real periodic 2d scalar data
   idrun = run identifier for current run, -1 if unknown
   """
   int_type = numpy.int32
   double_type = numpy.float64
   float_type = numpy.float32
   complex_type = numpy.complex64

   ns = 3
   iudm = 19; ius = 11
   nplot = 1

   dname = numpy.array(["POTENTIAL       ","ELECTRON DENSITY",
                        "ION DENSITY     "],dtype=str)

# create string from idrun
   if (idrun < 0):
      cdrun = "Unknown"
      while (cdrun.isdigit() == False):
         cdrun = input("enter integer idrun: ")
      idrun = int(cdrun)
   cdrun = str(idrun)
   fname = "diag2." + cdrun
   cmfield2.ffopen2(iudm,fname)

# nscalars = table of available diagnostics
   nscalars = numpy.zeros((ns),int_type,'F')

# determine which scalar diagnostics are available
   cmfield2.readsdiags2(iudm,nscalars)

   nts = numpy.zeros((1),int_type,'F')
   modesx = numpy.zeros((1),int_type,'F')
   modesy = numpy.zeros((1),int_type,'F')
   mrec = numpy.zeros((1),int_type,'F')
   fname = numpy.array([""],'S32')
   tname = numpy.array([""],'S32')
   ierr = numpy.zeros((1),int_type,'F')

# open graphics device
   ierr[0] = graf2.open_graphs2(nplot)
# set palette to color wheel
   graf2.set_palit(2)

# select diagnostic
   m = numpy.sum(nscalars)
   while True:
      if (m > 0):
         n = -1
         while True:
            if (n < 0):
               for i in range(0,ns):
                  if (nscalars[i]==1):
                     print ("enter ", i+1," for", 
                            numpy.str.rstrip(dname[i]))
               print ("enter ", 0," for EXIT")
               c = input("")
               if (c.isdigit()):
                  n = int(c)
               if (n==0):
                  break
               if ((n >= 1) and (n <= ns)):
                  if (nscalars[n-1]==0):
                     n = -1
            else:
               n = -1
            if (n > 0):
               break
            print ("invalid entry, try again or enter 0 to quit")
      else:
         print ("no scalar diagnostic files found")
         n = 0
# exit procedure
      if (n==0):
         if ("sfield" in globals()):
            sfield = None
         cmfield2.closeff2(iudm)
         graf2.close_graphs2()
         return

      tname = numpy.str.rstrip(dname[n-1])
      print (tname, " diagnostic selected")

# return parameters for selected scalar diagnostic:
# nts, modesx, modesy, nrec, fname
      cmfield2.sdiagparams2(iudm,n,nts,modesx,modesy,mrec,fname)
      nrec = mrec[0]

# nx/ny = number of global grid points in x/y direction
      nx = int(math.pow(2,in2.indx)); ny = int(math.pow(2,in2.indy))
# kyp = number of real grids in each field partition in y direction
      kyp = int((ny - 1)/in2.nvp) + 1
# kyb = minimum number of processors in distributed array
      kyb = int((ny - 1)/kyp) + 1
# nyv = second dimension of scalar field array, >= ny
      nyv = kyp*kyb

# allocate scalar array
      if ("sfield" not in globals()):
         sfield = numpy.empty((nx,nyv),float_type,'F')
      dt = in2.dt*float(nts[0])

# open stream file for scalar field
      cmfield2.fsopen2(ius,fname)

# nrec = number of complete records
      nrec = int(nrec/kyb)
      print ("records found: nrec = ", nrec)

# read and display data
      for ii in range(0,nrec):
# read real scalar field
         cmfield2.fread2(ius,sfield,nx,nyv)
         it = nts[0]*ii
         time = dt*float(ii)
# show time
         print ("it,time=",it,time)
# read and display data
         graf2.dscaler2(sfield,tname,it,999,0,1,nx,ny,ierr)
         if (ierr[0]==1):
            break

      cmfield2.closeff2(ius)
      print()

if (__name__=="__main__"):
   main(-1)
